### PR TITLE
Initial support for bfloat16

### PR DESCRIPTION
**Pull request by ansurya**
Initial support  is essentially clone of the rocBLAS bfloat16 header.
Added testcase to validate different bfloat16 operations on Host and Device.

### CODE REVIEW

File: include/hip/hcc_detail/hip_bfloat16.h [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    }
+    constexpr __host__ __device__ bool isnan(hip_bfloat16 a)
+    {
+        return !(~a.data & 0x7f80) && +(a.data & 0x7f);
+    }
+    constexpr __host__ __device__ bool iszero(hip_bfloat16 a)
+    {
+        return !(a.data & 0x7fff);
+    }
+    inline hip_bfloat16 sin(hip_bfloat16 a)
```
Comment (b-sumner) [inline include/hip/hcc_detail/hip_bfloat16.h]: We can remove these two math functions for now.

File: include/hip/hip_bfloat16.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+#ifndef HIP_INCLUDE_HIP_HIP_BFLOAT16_H
+#define HIP_INCLUDE_HIP_HIP_BFLOAT16_H
+
+#include <hip/hip_common.h>
+
+#if defined(__HIP_PLATFORM_HCC__) && !defined(__HIP_PLATFORM_NVCC__)
+#include <hip/hcc_detail/hip_bfloat16.h>
+#elif defined(__HIP_PLATFORM_NVCC__) && !defined(__HIP_PLATFORM_HCC__)
+#else
```
Comment (b-sumner) [inline include/hip/hip_bfloat16.h]: Can't we use the same implementation for NVCC?

File: tests/src/deviceLib/hipBfloat16.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+static std::mt19937 rng(dev());
+
+inline float getRandomFloat(float min = 10, float max = 230) {
+    std::uniform_real_distribution<float> gen(min, max);
+    return gen(rng);
+}
+
+__host__ __device__ bool testConversion(float a, hip_bfloat16 b) {
+  float c = float(b);
+  // float relative error should be less than 1/(2^7) since bfloat16
```
Comment (b-sumner) [inline tests/src/deviceLib/hipBfloat16.cpp]: This function is probably misnamed.  I would call it testRelativeAccuracy or something like that.

File: tests/src/deviceLib/hipBfloat16.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+#include <random>
+
+#define PI 3.14
+#define SIZE 100
+using namespace std;
+
+static std::random_device dev;
+static std::mt19937 rng(dev());
+
+inline float getRandomFloat(float min = 10, float max = 230) {
```
Comment (b-sumner) [inline tests/src/deviceLib/hipBfloat16.cpp]: I understand why this is avoiding zero, and I assume we need max-min != max, but why not make the interval bigger?

Comment (ansurya) [reply tests/src/deviceLib/hipBfloat16.cpp]: done.

**Comment by ansurya** [comment]
Test fails on Nvidia GPU, no development GPU s to debug the issue.

**Pull request merged**